[WIP]  new module ecs_tag

mpechner opened the pull request:
##### SUMMARYwrapper around boto3  ecs tag_resource() and untag_resource()
 <!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->

- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ecs_tag()  will tag ecs resources
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```
Started with ec2_tag module code.

abadger committed:
Add a sanity test for yaml to make sure we can parse all yaml

Check that all yaml we ship is parsable by the pyyaml c backend.  Since
Ansible uses Pyyaml for docs and playbooks, if the yaml files aren't
parsable, they will error out if they were used.

Warn and skip yamllint if libyaml backend is not present

Ignore new errors in examples until someone can fix them

abadger committed:
Change the check_pyyaml() function to return the features of pyyaml that we check

mariolenz committed:
VMware: Add advanced settings to vmware_cluster_ha

goneri committed:
prepare_vmware_tests: add scenario with one ESXi

Add the ability to test with just one vcenter instance and a single
ESXi. This allow us to cover 90 modules.

mpechner committed:
Wrapper around boto3 ecs tag_resource() and untag_resource()
Based on ansible module ec2_tag()

mpechner committed:
Wrapper around boto3 ecs tag_resource() and untag_resource()
Based on ansible module ec2_tag()

mpechner committed:
renamed ecs_tag

mpechner committed:
renamed ecs_tag

mpechner commented:
Just looking for comments on API.  Will write and check in tests  once I receive feedback.

pabelanger committed:
Split ios tests into network_cli and local connections (#65921)

This allows to group tests by connections, which allow us to disable
local connection testing in collections for the moment.

Signed-off-by: Paul Belanger <pabelanger@redhat.com>

samccann committed:
Use requirements.txt to install prereqs for docs build (#65709)

* Use requirements.txt to install prereqs for docs build

* implement feedback

joej164 committed:
Update aci_access_port_to_interface_policy_leaf_profile.py

##### SUMMARY
<!--- Your description here -->
The links in the file are wrong.  Updating the document name to point to the right document.

https://docs.ansible.com/ansible/latest/modules/aci_access_port_block_to_access_port_module.html

##### ISSUE TYPE
- Docs Pull Request

+label: docsite_pr

prasadkatti committed:
cloudformation integration tests (#65643)

tremble committed:
AnsibleAWSModule related cleanup: (#64931)

- fetch boto3 clients using module.client()
- remove ec2_argument_spec() use (AnsibleAWSModule adds it automatically)
- remove unused imports
- remove HAS_BOTO3 (AnsibleAWSModule handles it)

seanw122 committed:
Added ability to specify plan information for images that require it. (#65335)

* Modified to include plan information.

* Fixed where plan is passed to the virtual machine object.

* Added changelog file

* Update changelogs/fragments/65335-add-plan-to-azure-vmscaleset-module.yaml

Co-Authored-By: Felix Fontein <felix@fontein.de>

* Added plan suboptions with required flag. Removed code block checking plan since suboptions were added.

* Changed true to True. Added space after commas where failed tests indicated.

* Removed extra blank line. Added promotion_code to plan param list.

* Trying to fix indention issue

* Trying to fix indention

* Changed example capacity to trigger build check. Last failure was not due to code.

* Removed property for accepting terms and code block using it..

* Removed extra unneeded spaces.

mpechner committed:
Ansible test sanity checks are clean

mpechner committed:
Ansible test sanity checks are clean

mpechner committed:
corrected version_added to be 2.10

mpechner committed:
corrected version_added to be 2.10

mpechner commented:
I'll work on tests this weekend.  I wanted this to be reviewed, get feed back on the API I proposed. I am using a version locally  that only handles tagging the cluster resource.

dagwieers committed:
Use module references

goneri committed:
vmware/test: init esxi_user with esxi1_username

By default, `esxi_user` is initialize as `root`. We now use `esxi1_username`
instead.

Commit:
New zabbix_service module (#58926)

* New module zabbix_service (#5)

* new module zabbix_service

* fix type

* fix githubid

* New Zabbix service module (#11)

* new zabbix service module

* fix validate module failure

* Fix algorithm doc and all_childs param name

* Update Ansible version

Co-Authored-By: sky-joker <megane@kurobuti.com>

* remove dump state

tremble committed:
Add myself as a watcher on sanity/ignore.txt (#65950)

After spending the time and effort to clean out the AWS sanity failures I'd like to keep an eye open for new failures as the sanity tests are updated

borborpa committed:
Meraki: Fixed line 174 changed from None to the actual net_id… (#59395)

* Fixed line 174 changed from None to the actual net_id variable

* Update meraki_content_filtering.py

* Create 59395_meraki_content_filtering.yaml

pabelanger committed:
Add connection_local / connection_httpapi skip to eos (#65948)

This starts to allow us to break jobs out per connection testing.

Signed-off-by: Paul Belanger <pabelanger@redhat.com>

NilashishC committed:
Add debug (#65909)

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

waffie1 committed:
Add ChangeSets to cloudformation_info + updates (#64571)

* cloudformation_info updates

Add ability to retrieve stack change sets
Update module to use AnsibleAWSModule
Update backoff to use new decorator style
Updated exceptions to use fail_json_aws
Converted outputs to snake_case where appropriate, for cloudformation_info only to preserve cloudformation_facts functionality.
  Unconverted keys are stack_outputs, stack_parameters, stack_policy,
  stack_resources, stack_tags and stack_template

* cloudformation_info updates

Add ability to retrieve stack change sets
Update module to use AnsibleAWSModule
Update backoff to use new decorator style
Updated exceptions to use fail_json_aws
Converted outputs to snake_case where appropriate, for cloudformation_info only to preserve cloudformation_facts functionality.
  Unconverted keys are stack_outputs, stack_parameters, stack_policy,
  stack_resources, stack_tags and stack_template

* Adding integration tests

* Remove unneeded debug statement

* Add myself to Authors

* rename stack-name to match iam policy limits

aws-terminator permission for cloudformation have the following resource
Resource: arn:aws:cloudformation:us-east-1:{{ aws_account_id }}:stack/ansible-test*
updating test stack name to match

* removing .orig file from rebase

* fix stack name, underscore not allowed

* rename integration testsuite to cloudformation

rename per request

* add resource_prefix to resource names

* prefix stack name with ansible-test

IAM policy in aws-terminator requires the stack name to begin with ansible-test

* add single quotes around variable in test

* fix test for display name

mistakenly updated when changing topic_name to use resource_prefix.  The test is to check the display name, which is what the stack update changes

* fix value to display name test

* rename main.yaml to main.yml before rebase

* Merge tests with  #65643

remove uneeded files
using tests from #65643 with added check to validate changeset is present in info when requested

* fix assert on chageset check

* remove trailing blank line

tremble committed:
Run ec2_instance tests in parallel and re-enable them (#63800)

* Run ec2_instance tests in parallel

* Make sure we spin up the iam_instance_role instances in the dedicated VPC

* Rework a little to only build one VPC while still running in parallel

* Start running the tests within shippable

* We only use setup_remote_tmp_dir for the version_fail tests, so only set it up once

* Make sure we clean up if we fail during Environment Setup

* Add a warning about running in parallel

* Minor review tweaks

* Remove unused file

tremble committed:
ec2_launch_template: Add default_version and latest_version to output, matching the documentation (#61279)

- Add tests for default/latest version manipulation

haiyuazhang committed:
addd jie to azure dev (#65939)

roberbri committed:
Force values to be sorted when compared. (#65315)

* Force values to be sorted when compared.

This avoids problems where lists of values could be generated in
indeterminate order.

* Update lib/ansible/module_utils/aws/elbv2.py

Thank you felixfontein, sorry I missed this detail.

Co-Authored-By: Felix Fontein <felix@fontein.de>

felixfontein committed:
Add notes on docker_network connectivity for containers and services. (#65840)

billdodd committed:
Fix version param in deprecate() call and update porting guide (#65894)

* fix version= param in deprecate() call and update porting guide

* add changelog fragment

haiyuazhang committed:
add image reference in vmss instance info (#65872)

haiyuazhang committed:
azure_rm_virtualmachine: fixing issue #63844 (#65601)

mnecas committed:
ovirt_snapshot: select disks (#65729)

mnecas committed:
ovirt_network: correct external provider (#65701)

* ovirt_network: update external provider

* unnecessary remove blank line

* remove update check for external_provider

* first try to import then create

mnecas committed:
ovirt_affinity_group: add host rule option disabled (#65704)

Akasurde committed:
vars: Fail with warning for extra_vars filename without @ sign (#59915)

Rather than silently processing extra_vars filename without @ sign,
CLI now fails with appropriate warning about requirement.

Fixes: #51857

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

goneri committed:
vmware/test: more modular teardown.yml

Some teardown tasks are now enabled only if we have ESXi in the configuration
file. On my system and in a vcenter only scenario, `vcenter_folder` goes from
42.7 to 31.2s.

ivog74 committed:
Add ansible cli options --ask-vault-password and --vault-pass-file (#63782)

* Move new Ansible cli options '--ask-vault-password' and '--vault-pass-file' to the existing calls to add_argument
* Add changelog fragement
* Change order of ansible cli arguments to use --ask-vault-password and --vault-password-file by default
* Update runme.sh in vault integration tests to test new options --ask-vault-password and --vault-pass-file

internetionals committed:
Fix 'Assert CNAME failure' assertion (#65875)

waffie1 committed:
Cloudformation info checkmode (#65958)

* set supports_check_mode=True

I don't see any reason why this module cannot run in check mode.  The following API calls are made, none of which modify configurations

All using boto3.client('cloudformation')
describe_change_set
describe_stacks
describe_stack_events
get_paginator
get_stack_policy
get_template
list_change_sets
list_stack_resources

* duplicate cloudformation_info tasks with checkmode

Duplicated all existing cloudformation_info tests and added check_mode: yes to them

* delete duplicate empty line

SpamapS committed:
Add support for ECR Lifecycle Policies to ecs_ecr (#48997)

* Fix copy/pasta for ecs_ecr test names

* Add support for lifecycle policies to ecs_ecr

New feature for ecs_ecr to support [ECR Lifecycle Policies][].

Fixes #32003

 [ECR Lifecycle Policies]: https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html

* Improve error message for ecs_ecr parsing errors

Replaces the exception and stack trace with a description of what's
actually going wrong from a user perspective.

* Rename delete policy to purge policy

Marks the `delete_policy` parameter as deprecated, to be removed in
Ansible 2.6.

* Add version_added to purge_policy

* Remove changing results based on verbosity

What I really want is --diff support, and changing results based on
verbosity is abnormal.

* Ensure repository name is lowercase

* Fix deprecation cycle to 4 releases

* Use a YAML anchor for credentials

* Remove filters from assertions

* Add minimal permissions needed

* Updating version_added and deprecation cycle

The original PR sat while a few releases happened.

* Bumping version added and deprecation version

We missed the 2.8 release.

* Removing bare except:

This is not allowed and is generally bad practice.

* Fix lint errors

* update ansible release metadata

* Use the new alias deprecation scheme

This was added in the time the PR has been in development, so rework
things to use it.

* Add test coverage

This makes sure that lifecycle_policy is produced when passed in.

*Also a minor suggestion for simplification from PR.

* Restore changes from 62871 lost in rebase

* Add changelog

* Remove version_added for new purge_policy option

Per sanity test fail.

s-hertel committed:
free strategy - include failed hosts that were notified (#65576)

* free strategy - include failed hosts that were notified so --force-handlers is used

* trim line length a bit

* Loop over the force handler tests with the strategies linear and free

* rename changelog

* Use the play iterator instead of TQM for accurate failure representation in blocks

* Remove hack in a backwards compatible way for 3rd party plugins

bmillemathias committed:
Set a deprecation warning and version for option sha256sum (#65637)

The option sha256sum was declared deprecated more than 2 years ago without
setting a removal version. Let's do that now.

fixes #65636

bmillemathias committed:
[get_url] Return status_code on HTTP 304 (#65739)

Return field status_code on HTTP status 304 (implemented by lkthomas)
Add an integration test for this case.
Fixes #65263

mulatinho committed:
elb_{network_lb,target_group}: allow UDP/TCP_UDP (#65828)

elb_network_lb.py: allow UDP and TCP_UDP protocols

- Fixing documentation
- Add support to UDP and TCP_UDP as described on AWS SDK

elb_target_group.py: allow UDP, TLS, TCP_UDP proto

- Fixing documentation
- Making health checks with response codes and paths only in HTTP/HTTPS
- Allow UDP, TLS, TCP_UDP protocols as described on AWS SDK.

others:
- Added changelog fragments
- Integration test

Fixes: #65265
Signed-off-by: Alexandre Mulatinho <alex@mulatinho.net>

tremble committed:
fix ec2_vpc_net test instability (#65184)

* Move ec2_vpc_net tests to group1, group2 is running much longer

* Allow for VPC CIDRs to be "associat*ing*", things in AWS are "eventually consistent" and occasionally take longer than we would like

ngp-star committed:
VMware: filter VMs of Same name on the basis of folder (#65922)

vmware_guest now filters VMs of same name on the basis of additional information of folder.

sabre1041 committed:
VMware: Corrected documentation on vmware_guest module (#65733)

Update link for vapp_properties as per VMware docs in vmware_guest module

ganeshrn committed:
Fix netconf plugin related to collections (#65718)

* Fix netconf plugin related to collections

Fixes #65655 (partly)

*  Make netconf plugins configurable so that the
   information of ncclient device handler
   for give platform resides in the platform
   specific netconf plugin.
*  If the device handler value in ncclient is
   different from the ansible_network_os value
   the right value of `ncclient_device_handler`
   should be set in the plugin documentation.

* Fix review comments

* Fix CI issue

* Fix review comment

s-hertel committed:
Remove illusion of caching in the inventory script plugin (#64151)

* Load the cache plugin for the inventory script plugin and fix update cache logic

* Remove the illusion and let individual scripts deal with it

* reword that

* Deprecate instead of a hard failure

* changelog

mattclay committed:
Add ansible-test constraint to avoid coverage 5.0+

samdoran committed:
user - honor update_password parameter on BusyBox hosts (#65977)

The check for this parameter was missing from BusyBox.modify_user(), resulting in unexpected password changes to existing user accounts.

mattclay committed:
Disable failing azure_rm_storageaccount test.

pilou- committed:
clc_aa_policy: deprecate unused 'wait' parameter (#64367)

torvitas committed:
nmcli: Add compatibility for new networkmanager library (#65726)

Fixes: #48055
* add compatibility for new networkmanager library
* update debian/ubuntu docs
* move import gi into try

dano0b committed:
#58578 support absolute paths additionally (#58812)

* support absolute paths additionally

* add changelog

* rename changelog

* fix changelog to bugfix

dr4Ke committed:
terraform workspace list failure warns instead of fails (#65044)

* terraform workspace list failure warns instead of fails

Fixes #64788

* add a changelog fragment

atombrella committed:
Use context manager for file handling in misc files. (#65372)

The file handles are left open in those files.

stefanhorning committed:
Fix to return data when using lambda_info module (#64548)

* Fixed issue 64479 with lambda_info module

* Added integration tests for lambda_info module

* Moved lambda_info tests into already existing aws_lamda testsuite for easier test setup.

Co-authored-by: Jill R <4121322+jillr@users.noreply.github.com>

D3DeFi committed:
zabbix_hostmacro support for zabbix native macro names and macro_value not required when state=absent (#65884)

mpechner committed:
Made tags optional if state is list
For remove tags, list of keys has to be a list or tupel

Tested:
all 3 states for cluster, service and task_defintion.
No Ansible info for tasks and cluster_instances - or did I miss something. So these resources types have not been tested.

mpechner committed:
Made tags optional if state is list
For remove tags, list of keys has to be a list or tupel

Tested:
all 3 states for cluster, service and task_defintion.
No Ansible info for tasks and cluster_instances - or did I miss something. So these resources types have not been tested.

mpechner commented:
How do I get someone to tag a look? next steps?  Do you want the Ansible playbooks I used to test placed somewhere?

 Made tags optional if state is list
    For remove tags, list of keys has to be a list or tupel
    
    Tested:
    all 3 states for cluster, service and task_defintion.
    No Ansible info for tasks and cluster_instances - or did I miss something. So these resources types have not been tested.

milouse committed:
Don't rely anymore on locale-dependent wording in Pacman output (#65878)

Use line position instead, as specified in upstream pacman source code.

See: https://git.archlinux.org/pacman.git/tree/src/pacman/package.c#n260
Fixes: https://github.com/ansible/ansible/issues/64682
Fixes: https://github.com/ansible/ansible/issues/65237

Akasurde committed:
Revert "Don't rely anymore on locale-dependent wording in Pacman output (#65878)" (#66029)

This reverts commit cd0e038b746f426003ba7177f5ff4882db9dfe9b.

sky-joker committed:
Include groups in name-based search results (#66026)

add changelogs

L3n41c committed:
Fix pacman stdout parsing in the Ansible module (#65238)

pacman output is localized and the Ansible module is parsing its output.
So, we need to force the locale.
Add changelog fragment

Fixes #65237

evgeni committed:
don't use `with_file` in exclusive=True examples (#65967)

geerlingguy committed:
Fixes #65952: Remove ambiguous example from k8s docs to clarify file usage. (#65953)

Commit:
fix AWS cloudfront purge_cache_behaviors misspelling (#65927)

tpraxl committed:
Be more specific about the requirements per distro (#65810)

D3DeFi committed:
improved zabbix action conditions documentation in zabbix_action module (#65808)

choeffer committed:
Update os_group_info.py (#65115)

abadger committed:
Add a banner which will display on the testing site. (#64763)

* Add a banner which will display on the testing site.

Thanks to shane for the javascript code

* change to more noticable color

Co-authored-by: Sandra McCann <samccann@redhat.com>

Akasurde committed:
VMware: Handle NoneType in vmware_vmkernel_info (#65834)

Check for virtualNicManager in Esxi host system before accessing properties in vmware_vmkernel_info.

Fixes: #62772

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Check if task file is specified for import_tasks (#57572)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

trishnaguha committed:
fix nxos_hsrp keyerror (#65796)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

haiyuazhang committed:
fixing #64404 (#66032)

NilashishC committed:
Fix nxos_l3_interfaces rendering of dhcp (#66049)

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

Akasurde committed:
VMware: Add example for creating VM without disks (#66057)

Fixes: #39743

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

abhaykadam committed:
Fix env lookup plugin error on utf8 values (#65541)

This commit fixes issue #65297.

The env lookup plugin used to fail when environment variable value
contained any UTF-8 characters (e.g., δ, ζ).

D3DeFi committed:
zabbix_host idempotency fix for situation where ZabbixAPI handles host groups case insensitively (#63860)

Akasurde committed:
homebrew_cask: Update documentation and examples (#66073)

* properly document ``upgrade`` option as an alias to ``upgrade_all``, and not a separate option
* Update documentation.
* Update examples.

Fixes: #34696

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

pypypyo14 committed:
Homebrew: some functions ignored check_mode option (#65387)

update_homebrew and upgrade_all now support check_mode

Fixes: #65352

Commit:
os_volume: optional bootable parameter for volume (#43945)

rohitthakur2590 committed:
allowing banner without stripping spaces (#62573)

Signed-off-by: rohitthakur2590 <rohitthakur2590@outlook.com>

mikewiebe committed:
Fix nxos_user roles bug (#65962)

ezmac commented:
Ansibot should have mentioned IRC a few times.  If you want to talk to someone about this PR, that's the best place to go.  irc.freenode.net #ansible-aws Tests will help but from what I've seen, you'll need to talk to others on IRC.  Sometimes it can take a while for them to get the time to look.
I'm "on holiday" until jan-6, but I've been watching the original issue for a while.  I'm just community, but I'll try to take a look when I get back.

andreaso committed:
nsupdate: Use provided TSIG key for all queries (#63174)

In addition to signing update queries also use the TSIG key to sign
lookup queries. By doing that we allow a hidden master to not only to
be looked down network wise, but also TSIG wise.

A bonus benefit of threating update queries and lookup queries more
the same is that will allow for all queries to be refactored into a
shared helper method. Currently we have a bit too much duplicated code
within the module.

justjais committed:
To fix the bug where IOS vlans module throws traceback (#64633)

* fix 64515

* add fix

mpechner commented:
Thanks.  Posted a message on #ansible-aws and joined ansible-devel and ansible

--------------------------------------------
Michael Pechner  NE6RD

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Friday, December 27, 2019 6:43 AM, Tad Merchant <notifications@github.com> wrote:

> Ansibot should have mentioned IRC a few times. If you want to talk to someone about this PR, that's the best place to go. irc.freenode.net #ansible-aws Tests will help but from what I've seen, you'll need to talk to others on IRC. Sometimes it can take a while for them to get the time to look.
> I'm "on holiday" until jan-6, but I've been watching the original issue for a while. I'm just community, but I'll try to take a look when I get back.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/ansible/ansible/pull/65924?email_source=notifications&email_token=AABTQSCVYXW2SY4KN2CENNLQ2YIB5A5CNFSM4J37IEJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXJGLA#issuecomment-569283372), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AABTQSH67DKYWAY4NDF23T3Q2YIB5ANCNFSM4J37IEJA).

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+  resource:
+    description:
+      - The ecs resource name. Can be none only if resource_type is cluster.
+    required: false
```
```suggestion
```

By convention we don't add ``required: false`` (it's implied) See also
https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+  tags:
+    description:
+      - A dictionary of tags to add or remove from the resource.
+      - If the value provided for a tag is null and C(state) is I(absent), the tag will be removed regardless of its current value.
```
```suggestion
      - If the value provided for a tag is null and I(state=absent), the tag will be removed regardless of its current value.
```

C is for values I is for parameters (and can be used for param=) there are some bad examples around.

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+  purge_tags:
+    description:
+      - Whether unspecified tags should be removed from the resource.
+      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
```
```suggestion
      - Note that when combined with I(state=absent), specified tags with non-matching values are not purged.
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+      - Whether unspecified tags should be removed from the resource.
+      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
+    type: bool
+    default: no
```
```suggestion
    default: false
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+DOCUMENTATION = r'''
+---
+module: ecs_tag
+short_description: create and remove tags on ecs clusters.
```
```suggestion
short_description: create and remove tags on Amazon ECS resources
```

a) short_description shouldn't end in a '.'
b) ECS is an acronym and product name.  In the documentation we should try to refer to it as 'ECS' rather than 'ecs' as an accuracy thing.
c) If I'm reading the module correctly it doesn't just tag clusters...

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+notes:
+    - none
+description:
+    - Creates, removes and lists tags for an ecs_cluster.
```
```suggestion
    - Creates, removes and lists tags for Amazon ECS resources.
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+    - none
+description:
+    - Creates, removes and lists tags for an ecs_cluster.
+       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
```
```suggestion
     - Resources are referenced by their cluster name.
```

I wouldn't bother with the comment about 'complex args', the parameters and examples do this for you.

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+    - Creates, removes and lists tags for an ecs_cluster.
+       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
+version_added: '2.10'
+author:
```
@acozine : This is based on the ec2_tags module written by @lwade and @flowerysong.  It feels 'wrong' adding them as authors here since this will start spamming them for ecs_tags issues/PRs.

Is there any 'standard' approach here?  Put a note below the copyright notice?

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+options:
+  cluster_name:
+    description:
+      - The name of the cluster who's resources we are tagging
```
```suggestion
      - The name of the cluster who's resources we are tagging.
```

While short_description entries shouldn't have a '.' at the end, description entries should...

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+
+try:
+    from botocore.exceptions import BotoCoreError, ClientError
+except Exception:
```
```suggestion
except ImportError:
```

It's generally considered a 'bad' thing to catch 'Exception'.  This is normally done by catching ImportError...

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+        resource_type=dict(default='cluster', choices=['cluster', 'task', 'service', 'task_definition', 'container'])
+
+    )
+    required_if = [('state', 'present', ['tags']), ('state', 'absent', ['tags'])]
```
I think the code below is going to spit out an ugly exception if resource_type is task/service/task_definition/container and resource is undefined.

It's probably better to require resource be set if resource_type is one of those values...

```suggestion
    required_if = [
        ('state', 'present', ['tags']),
        ('state', 'absent', ['tags']),
        ('resource_type', 'task', ['resource']),
        ('resource_type', 'service', ['resource']),
        ('resource_type', 'task_definition', ['resource']),
        ('resource_type', 'container', ['resource']),
    ]
```

(note: I've not double checked the spacing here, so it'll possibly generate a lint warning)

tremble reviewed (changes requested):
Hi,

Thanks for your contribution.  Mostly little niggles, I know you lifted a lot of the work from ec2_tag but we're trying to raise the quality of our code so you're going to get a little more scrutiny than the original code faced.  (We are going back and cleaning things up, but this takes time)

In addition to my other comments, running the code through pylint (with complexity testing enabled) is picking up on the main function getting a little 'complex':  "The McCabe rating is 27 (too-complex)".  While this doesn't generate a sanity error today there is work underway to add complexity testing to the sanity tests.  ( #64623 )

I'd suggest pulling the code for finding the ARN out into a dedicated function.  Additionally, rather than the mass of ``if len(...):`` clauses you can just directly do the resource_arn assignments and catch Index/Key errors (just after catching the Boto errors) if no resource was returned.

```
    try:
        if resource_type == 'cluster':
            description = ecs.describe_clusters(clusters=[resource])
            resource_arn = description['clusters'][0]['clusterArn']
        elif resource_type == 'task':
...
    except (IndexError, KeyError):
        module.fail_json(msg='Failed to find {0} {1}'.format(resource_type, resource))
```

Also, as mentioned on IRC, please add some integration tests and update lib/ansible/config/module_defaults.yml

mpechner commented:
I should have time to get this done this weekend. Thanks for the feedback.

--------------------------------------------
Michael Pechner  NE6RD

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Friday, December 27, 2019 11:17 AM, Mark Chappell <notifications@github.com> wrote:

> @tremble requested changes on this pull request.
>
> Hi,
>
> Thanks for your contribution. Mostly little niggles, I know you lifted a lot of the work from ec2_tag but we're trying to raise the quality of our code so you're going to get a little more scrutiny than the original code faced. (We are going back and cleaning things up, but this takes time)
>
> In addition to my other comments, running the code through pylint (with complexity testing enabled) is picking up on the main function getting a little 'complex': "The McCabe rating is 27 (too-complex)". While this doesn't generate a sanity error today there is work underway to add complexity testing to the sanity tests. ( [#64623](https://github.com/ansible/ansible/pull/64623) )
>
> I'd suggest pulling the code for finding the ARN out into a dedicated function. Additionally, rather than the mass of if len(...): clauses you can just directly do the resource_arn assignments and catch Index/Key errors (just after catching the Boto errors) if no resource was returned.
>
> except (IndexError, KeyError) as e:
>         module.fail_json(msg='Failed to find {0} {1}'.format(resource_type, resource))
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361710166):
>
>> +version_added: '2.10'
> +author:
> +  - Lester Wade (@lwade)
> +  - Paul Arthur (@flowerysong)
> +  - Michael Pechner (@mpechner)
> +requirements: [ boto3, botocore ]
> +options:
> +  cluster_name:
> +    description:
> +      - The name of the cluster who's resources we are tagging
> +    required: true
> +    type: str
> +  resource:
> +    description:
> +      - The ecs resource name. Can be none only if resource_type is cluster.
> +    required: false
>
> ⬇️ Suggested change
>
> -    required: false
>
> By convention we don't add required: false (it's implied) See also
> https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361710959):
>
>> +  resource_type:
> +    description:
> +      - The type of resource
> +    default: cluster
> +    choices: ['cluster', 'task', 'service', 'task_definition', 'container']
> +    type: str
> +  state:
> +    description:
> +      - Whether the tags should be present or absent on the resource. Use list to interrogate the tags of a ecs resource.
> +    default: present
> +    choices: ['present', 'absent', 'list']
> +    type: str
> +  tags:
> +    description:
> +      - A dictionary of tags to add or remove from the resource.
> +      - If the value provided for a tag is null and C(state) is I(absent), the tag will be removed regardless of its current value.
>
> ⬇️ Suggested change
>
> -      - If the value provided for a tag is null and C(state) is I(absent), the tag will be removed regardless of its current value.
> +      - If the value provided for a tag is null and I(state=absent), the tag will be removed regardless of its current value.
>
> C is for values I is for parameters (and can be used for param=) there are some bad examples around.
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361711227):
>
>> +    type: str
> +  state:
> +    description:
> +      - Whether the tags should be present or absent on the resource. Use list to interrogate the tags of a ecs resource.
> +    default: present
> +    choices: ['present', 'absent', 'list']
> +    type: str
> +  tags:
> +    description:
> +      - A dictionary of tags to add or remove from the resource.
> +      - If the value provided for a tag is null and C(state) is I(absent), the tag will be removed regardless of its current value.
> +    type: dict
> +  purge_tags:
> +    description:
> +      - Whether unspecified tags should be removed from the resource.
> +      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
>
> ⬇️ Suggested change
>
> -      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
> +      - Note that when combined with I(state=absent), specified tags with non-matching values are not purged.
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361711277):
>
>> +    description:
> +      - Whether the tags should be present or absent on the resource. Use list to interrogate the tags of a ecs resource.
> +    default: present
> +    choices: ['present', 'absent', 'list']
> +    type: str
> +  tags:
> +    description:
> +      - A dictionary of tags to add or remove from the resource.
> +      - If the value provided for a tag is null and C(state) is I(absent), the tag will be removed regardless of its current value.
> +    type: dict
> +  purge_tags:
> +    description:
> +      - Whether unspecified tags should be removed from the resource.
> +      - "Note that when combined with C(state: absent), specified tags with non-matching values are not purged."
> +    type: bool
> +    default: no
>
> ⬇️ Suggested change
>
> -    default: no
> +    default: false
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361714488):
>
>> @@ -0,0 +1,238 @@
> +#!/usr/bin/python
> +# -*- coding: utf-8 -*-
> +# Copyright: (c) 2019, Michael Pechner <mikey@mikey.com>
> +# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
> +from __future__ import absolute_import, division, print_function
> +ANSIBLE_METADATA = {'metadata_version': '1.1',
> +                    'status': ['preview'],
> +                    'supported_by': 'community'}
> +
> +
> +DOCUMENTATION = r'''
> +---
> +module: ecs_tag
> +short_description: create and remove tags on ecs clusters.
>
> ⬇️ Suggested change
>
> -short_description: create and remove tags on ecs clusters.
> +short_description: create and remove tags on Amazon ECS resources
>
> a) short_description shouldn't end in a '.'
> b) ECS is an acronym and product name. In the documentation we should try to refer to it as 'ECS' rather than 'ecs' as an accuracy thing.
> c) If I'm reading the module correctly it doesn't just tag clusters...
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361714746):
>
>> +# Copyright: (c) 2019, Michael Pechner <mikey@mikey.com>
> +# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
> +from __future__ import absolute_import, division, print_function
> +ANSIBLE_METADATA = {'metadata_version': '1.1',
> +                    'status': ['preview'],
> +                    'supported_by': 'community'}
> +
> +
> +DOCUMENTATION = r'''
> +---
> +module: ecs_tag
> +short_description: create and remove tags on ecs clusters.
> +notes:
> +    - none
> +description:
> +    - Creates, removes and lists tags for an ecs_cluster.
>
> ⬇️ Suggested change
>
> -    - Creates, removes and lists tags for an ecs_cluster.
> +    - Creates, removes and lists tags for Amazon ECS resources.
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361715102):
>
>> +# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
> +from __future__ import absolute_import, division, print_function
> +ANSIBLE_METADATA = {'metadata_version': '1.1',
> +                    'status': ['preview'],
> +                    'supported_by': 'community'}
> +
> +
> +DOCUMENTATION = r'''
> +---
> +module: ecs_tag
> +short_description: create and remove tags on ecs clusters.
> +notes:
> +    - none
> +description:
> +    - Creates, removes and lists tags for an ecs_cluster.
> +       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
>
> ⬇️ Suggested change
>
> -       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
> +     - Resources are referenced by their cluster name.
>
> I wouldn't bother with the comment about 'complex args', the parameters and examples do this for you.
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361716093):
>
>> +ANSIBLE_METADATA = {'metadata_version': '1.1',
> +                    'status': ['preview'],
> +                    'supported_by': 'community'}
> +
> +
> +DOCUMENTATION = r'''
> +---
> +module: ecs_tag
> +short_description: create and remove tags on ecs clusters.
> +notes:
> +    - none
> +description:
> +    - Creates, removes and lists tags for an ecs_cluster.
> +       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
> +version_added: '2.10'
> +author:
>
> [@acozine](https://github.com/acozine) : This is based on the ec2_tags module written by [@lwade](https://github.com/lwade) and [@flowerysong](https://github.com/flowerysong). It feels 'wrong' adding them as authors here since this will start spamming them for ecs_tags issues/PRs.
>
> Is there any 'standard' approach here? Put a note below the copyright notice?
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361716318):
>
>> +short_description: create and remove tags on ecs clusters.
> +notes:
> +    - none
> +description:
> +    - Creates, removes and lists tags for an ecs_cluster.
> +       The resource is referenced by its cluster name.  It is designed to be used with complex args (tags), see the examples.
> +version_added: '2.10'
> +author:
> +  - Lester Wade (@lwade)
> +  - Paul Arthur (@flowerysong)
> +  - Michael Pechner (@mpechner)
> +requirements: [ boto3, botocore ]
> +options:
> +  cluster_name:
> +    description:
> +      - The name of the cluster who's resources we are tagging
>
> ⬇️ Suggested change
>
> -      - The name of the cluster who's resources we are tagging
> +      - The name of the cluster who's resources we are tagging.
>
> While short_description entries shouldn't have a '.' at the end, description entries should...
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361717154):
>
>> +added_tags:
> +  description: A dict of tags that were added to the resource
> +  returned: If tags were added
> +  type: dict
> +removed_tags:
> +  description: A dict of tags that were removed from the resource
> +  returned: If tags were removed
> +  type: dict
> +'''
> +
> +from ansible.module_utils.aws.core import AnsibleAWSModule
> +from ansible.module_utils.ec2 import boto3_tag_list_to_ansible_dict, ansible_dict_to_boto3_tag_list, compare_aws_tags
> +
> +try:
> +    from botocore.exceptions import BotoCoreError, ClientError
> +except Exception:
>
> ⬇️ Suggested change
>
> -except Exception:
> +except ImportError:
>
> It's generally considered a 'bad' thing to catch 'Exception'. This is normally done by catching ImportError...
>
> ---------------------------------------------------------------
>
> In [lib/ansible/modules/cloud/amazon/ecs_tag.py](https://github.com/ansible/ansible/pull/65924#discussion_r361723689):
>
>> +        return boto3_tag_list_to_ansible_dict(ecs.list_tags_for_resource(resourceArn=resource)['tags'])
> +    except (BotoCoreError, ClientError) as e:
> +        module.fail_json_aws(e, msg='Failed to fetch tags for resource {0}'.format(resource))
> +
> +
> +def main():
> +    argument_spec = dict(
> +        cluster_name=dict(required=True),
> +        resource=dict(required=False),
> +        tags=dict(type='dict'),
> +        purge_tags=dict(type='bool', default=False),
> +        state=dict(default='present', choices=['present', 'absent', 'list']),
> +        resource_type=dict(default='cluster', choices=['cluster', 'task', 'service', 'task_definition', 'container'])
> +
> +    )
> +    required_if = [('state', 'present', ['tags']), ('state', 'absent', ['tags'])]
>
> I think the code below is going to spit out an ugly exceptin if resource_type is task/service/task_definition/container and resource is undefined.
>
> It's probably better to require resource be set if resource_type is one of those values...
>
> ⬇️ Suggested change
>
> -    required_if = [('state', 'present', ['tags']), ('state', 'absent', ['tags'])]
> +    required_if = [
> +        ('state', 'present', ['tags']),
> +        ('state', 'absent', ['tags']),
> +        ('resource_type', 'task', ['resource']),
> +        ('resource_type', 'service', ['resource']),
> +        ('resource_type', 'task_definition', ['resource']),
> +        ('resource_type', 'container', ['resource']),
> +    ]
>
> (note: I've not double checked the spacing here, so it'll possible generate a lint warning)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/ansible/ansible/pull/65924?email_source=notifications&email_token=AABTQSG5ADODKWQJU2UUNZ3Q2ZIDNA5CNFSM4J37IEJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQJZNKY#pullrequestreview-336828075), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AABTQSBGNOBBKZSVLOWIP33Q2ZIDNANCNFSM4J37IEJA).

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

remove unneeded setting

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

remove unneeded setting

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

false, not no

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

false, not no

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

I like the comment change

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

I like the comment change

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

corrected style

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

corrected style

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

thanks for the specific

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

thanks for the specific

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

thanks for the terminology fix

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

thanks for the terminology fix

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Correct, supposed to tag any ECS resource

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Correct, supposed to tag any ECS resource

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,237 @@
+    type: str
+  resource:
+    description:
+      - The ecs resource name. Can be none only if resource_type is cluster.
```
```suggestion
      - The ECS resource name.
      - Required unless I(resource_type=cluster).
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,237 @@
+    - none
+description:
+    - Creates, removes and lists tags for Amazon ECS resources.
+     - Resources are referenced by their cluster name.
```
```suggestion
    - Resources are referenced by their cluster name.
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,237 @@
+    type: str
+  resource_type:
+    description:
+      - The type of resource
```
```suggestion
      - The type of resource.
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,237 @@
+    type: str
+  state:
+    description:
+      - Whether the tags should be present or absent on the resource. Use list to interrogate the tags of a ecs resource.
```
```suggestion
      - Whether the tags should be present or absent on the resource.
      - Use C(list) to interrogate the tags of an ECS resource.
```

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,237 @@
+options:
+  cluster_name:
+    description:
+      - The name of the cluster who's resources we are tagging.
```
```suggestion
      - The name of the cluster whose resources we are tagging.
```

tremble reviewed (changes requested):
Some more minor documentation tweaks.

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

mpechner committed:
Update lib/ansible/modules/cloud/amazon/ecs_tag.py

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

Brajen259 committed:
update aws_eks_cluster.py (#66105)

palvarez89 committed:
docker_swarm_service: Fix description of limits.memory (#66078)

This option corresponds to the '--limit-memory' option. Also
fix other mentions memory reservation instead of limit.

felixfontein committed:
docker_network: fix multiple subnet (of same IP version) idempotence (#65839)

* Fix multiple subnet (of same IP version) idempotence for docker_network.

* Add changelog.

* Unit tests no longer make sense, since the part of the code they test has been removed.

* Re-add CIDR validation. Move it to better position (module setup instead of idempotence check).

* Update changelog.

* Only run new tests on VM test images.

* Actually do what is documented. Especially since an empty object is a valid value for aux_addresses.

felixfontein committed:
docker_container: wait for removal if removal is in process (#65854)

* Allow to inspect containers directly.

* Wait for containers to be removed before recreating them.

* Also wait for containers to be removed before creating them.

* Add changelog.

mpechner committed:
Merge branch 'devel' of github.com:mpechner/ansible into devel

mpechner commented:
When  a rebase does not work, stop and try again.
